// @linkurious/rest-client/dist/src/api/CustomAction/index.js
"use strict";
/**
 * Copyright Linkurious SAS 2012 - 2019
 *
 * - Created on 2019-10-24.
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var request_1 = require("../../http/request");
var response_1 = require("../../http/response");
__export(require("./types"));
var UNAUTHORIZED = response_1.LkErrorKey.UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE = response_1.LkErrorKey.DATA_SOURCE_UNAVAILABLE, GUEST_DISABLED = response_1.LkErrorKey.GUEST_DISABLED, FORBIDDEN = response_1.LkErrorKey.FORBIDDEN, NOT_FOUND = response_1.LkErrorKey.NOT_FOUND, MALFORMED_CUSTOM_ACTION_TEMPLATE = response_1.LkErrorKey.MALFORMED_CUSTOM_ACTION_TEMPLATE;
var CustomActionAPI = /** @class */ (function (_super) {
    __extends(CustomActionAPI, _super);
    function CustomActionAPI() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Get all the custom actions owned by the current user or shared with them.
     */
    CustomActionAPI.prototype.getCustomActions = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, GUEST_DISABLED, FORBIDDEN],
            url: '/:sourceKey/customAction',
            method: 'GET',
            params: params
        });
    };
    /**
     * Create a new custom action for the current user.
     */
    CustomActionAPI.prototype.createCustomAction = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, FORBIDDEN, MALFORMED_CUSTOM_ACTION_TEMPLATE],
            url: '/:sourceKey/customAction',
            method: 'POST',
            params: params
        });
    };
    /**
     * Update a custom action owned by the current user.
     */
    CustomActionAPI.prototype.updateCustomAction = function (params) {
        return this.request({
            errors: [
                UNAUTHORIZED,
                DATA_SOURCE_UNAVAILABLE,
                FORBIDDEN,
                NOT_FOUND,
                MALFORMED_CUSTOM_ACTION_TEMPLATE
            ],
            url: '/:sourceKey/customAction/:id',
            method: 'PATCH',
            params: params
        });
    };
    /**
     * Delete a custom action owned by the current user.
     */
    CustomActionAPI.prototype.deleteCustomAction = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, FORBIDDEN, NOT_FOUND],
            url: '/:sourceKey/customAction/:id',
            method: 'DELETE',
            params: params
        });
    };
    return CustomActionAPI;
}(request_1.Request));
exports.CustomActionAPI = CustomActionAPI;
