// @linkurious/rest-client/dist/src/api/CustomAction/types.js
"use strict";
/**
 * Copyright Linkurious SAS 2012 - 2019
 *
 * - Created on 2019-08-19.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var CustomActionType;
(function (CustomActionType) {
    CustomActionType["NON_GRAPH"] = "non-graph";
    CustomActionType["NODE"] = "node";
    CustomActionType["EDGE"] = "edge";
    CustomActionType["NODESET"] = "nodeset";
    CustomActionType["EDGESET"] = "edgeset";
})(CustomActionType = exports.CustomActionType || (exports.CustomActionType = {}));
var CustomActionVariable;
(function (CustomActionVariable) {
    CustomActionVariable["BASE_URL"] = "baseurl";
    CustomActionVariable["VISUALIZATION"] = "visualization";
    CustomActionVariable["SOURCE_KEY"] = "sourcekey";
    CustomActionVariable["NODE"] = "node";
    CustomActionVariable["EDGE"] = "edge";
    CustomActionVariable["NODE_SET"] = "nodeset";
    CustomActionVariable["EDGE_SET"] = "edgeset";
})(CustomActionVariable = exports.CustomActionVariable || (exports.CustomActionVariable = {}));
var CustomActionSharingMode;
(function (CustomActionSharingMode) {
    CustomActionSharingMode["PRIVATE"] = "private";
    CustomActionSharingMode["SOURCE"] = "source";
})(CustomActionSharingMode = exports.CustomActionSharingMode || (exports.CustomActionSharingMode = {}));
var CustomActionRight;
(function (CustomActionRight) {
    CustomActionRight["OWNER"] = "owner";
    CustomActionRight["READ"] = "read";
})(CustomActionRight = exports.CustomActionRight || (exports.CustomActionRight = {}));
var CustomActionParsingErrorKey;
(function (CustomActionParsingErrorKey) {
    CustomActionParsingErrorKey["UNCLOSED_EXPRESSION"] = "unclosed-expression";
    CustomActionParsingErrorKey["EMPTY_EXPRESSION"] = "empty-expression";
    CustomActionParsingErrorKey["INVALID_EXPRESSION_SYNTAX"] = "invalid-expression-syntax";
    CustomActionParsingErrorKey["INVALID_VARIABLE"] = "invalid-variable";
    CustomActionParsingErrorKey["INVALID_SEMANTIC"] = "invalid-semantic";
    CustomActionParsingErrorKey["UNKNOWN_NODE_CATEGORY"] = "unknown-node-category";
    CustomActionParsingErrorKey["UNKNOWN_EDGE_TYPE"] = "unknown-edge-type";
    CustomActionParsingErrorKey["NO_EXPRESSIONS"] = "no-expressions";
    CustomActionParsingErrorKey["INVALID_TEMPLATE_COMBINATION"] = "invalid-template-combination";
    CustomActionParsingErrorKey["INCOMPATIBLE_RESTRICTIONS"] = "incompatible-restrictions";
})(CustomActionParsingErrorKey = exports.CustomActionParsingErrorKey || (exports.CustomActionParsingErrorKey = {}));
