// @linkurious/rest-client/dist/src/api/GraphEdge/index.js
"use strict";
/**
 * Copyright Linkurious SAS 2012 - 2019
 *
 * - Created on 2019-10-30.
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var request_1 = require("../../http/request");
var response_1 = require("../../http/response");
var UNAUTHORIZED = response_1.LkErrorKey.UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE = response_1.LkErrorKey.DATA_SOURCE_UNAVAILABLE, GUEST_DISABLED = response_1.LkErrorKey.GUEST_DISABLED, FORBIDDEN = response_1.LkErrorKey.FORBIDDEN, NOT_FOUND = response_1.LkErrorKey.NOT_FOUND, EDIT_CONFLICT = response_1.LkErrorKey.EDIT_CONFLICT;
var GraphEdgeAPI = /** @class */ (function (_super) {
    __extends(GraphEdgeAPI, _super);
    function GraphEdgeAPI() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Get an edge of the graph.
     * A subgraph made of the edge and its extremities is returned.
     */
    GraphEdgeAPI.prototype.getEdge = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, GUEST_DISABLED, NOT_FOUND],
            url: '/:sourceKey/graph/edges/:id',
            method: 'POST',
            params: params
        });
    };
    /**
     * Add an edge to the graph.
     */
    GraphEdgeAPI.prototype.createEdge = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, FORBIDDEN],
            url: '/:sourceKey/graph/edges',
            method: 'POST',
            params: params
        });
    };
    /**
     * Update a subset of properties of an edge.
     * Keep every other property of the edge unchanged.
     * It's not possible to update the type of an edge.
     */
    GraphEdgeAPI.prototype.updateEdge = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, FORBIDDEN, NOT_FOUND, EDIT_CONFLICT],
            url: '/:sourceKey/graph/edges/:id',
            method: 'PATCH',
            params: params
        });
    };
    /**
     * Delete an edge from the graph.
     */
    GraphEdgeAPI.prototype.deleteEdge = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, FORBIDDEN, NOT_FOUND],
            url: '/:sourceKey/graph/edges/:id',
            method: 'DELETE',
            params: params
        });
    };
    /**
     * Get the number of edges in the graph.
     */
    GraphEdgeAPI.prototype.getEdgeCount = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, GUEST_DISABLED],
            url: '/:sourceKey/graph/edges/count',
            method: 'GET',
            params: params
        });
    };
    return GraphEdgeAPI;
}(request_1.Request));
exports.GraphEdgeAPI = GraphEdgeAPI;
