// @linkurious/rest-client/dist/src/api/Config/types.js
"use strict";
/**
 * Copyright Linkurious SAS 2012 - 2019
 *
 * - Created on 2019-08-19.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var DefaultPage;
(function (DefaultPage) {
    DefaultPage["DASHBOARD"] = "dashboard";
    DefaultPage["WORKSPACE"] = "workspace";
})(DefaultPage = exports.DefaultPage || (exports.DefaultPage = {}));
var AntiAliasing;
(function (AntiAliasing) {
    AntiAliasing["SUPER_SAMPLING"] = "super-sampling";
    AntiAliasing["NATIVE"] = "native";
    AntiAliasing["NONE"] = "none";
})(AntiAliasing = exports.AntiAliasing || (exports.AntiAliasing = {}));
var OgmaRenderer;
(function (OgmaRenderer) {
    OgmaRenderer["WEBGL"] = "webgl";
    OgmaRenderer["CANVAS"] = "canvas";
})(OgmaRenderer = exports.OgmaRenderer || (exports.OgmaRenderer = {}));
var ImgCrossOrigin;
(function (ImgCrossOrigin) {
    ImgCrossOrigin["ANONYMOUS"] = "anonymous";
    ImgCrossOrigin["USE_CREDENTIALS"] = "use-credentials";
})(ImgCrossOrigin = exports.ImgCrossOrigin || (exports.ImgCrossOrigin = {}));
