// @linkurious/rest-client/dist/src/api/displayTypes.js
"use strict";
/**
 * LINKURIOUS CONFIDENTIAL
 * Copyright Linkurious SAS 2012 - 2019
 *
 * - Created on 2019-11-05.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var SelectorType;
(function (SelectorType) {
    SelectorType["ANY"] = "any";
    SelectorType["NO_VALUE"] = "novalue";
    SelectorType["NAN"] = "nan";
    SelectorType["RANGE"] = "range";
    SelectorType["IS"] = "is";
})(SelectorType = exports.SelectorType || (exports.SelectorType = {}));
var OgmaNodeShape;
(function (OgmaNodeShape) {
    OgmaNodeShape["CIRCLE"] = "circle";
    OgmaNodeShape["CROSS"] = "cross";
    OgmaNodeShape["DIAMOND"] = "diamond";
    OgmaNodeShape["PENTAGON"] = "pentagon";
    OgmaNodeShape["SQUARE"] = "square";
    OgmaNodeShape["STAR"] = "star";
    OgmaNodeShape["EQUILATERAL"] = "equilateral";
})(OgmaNodeShape = exports.OgmaNodeShape || (exports.OgmaNodeShape = {}));
var OgmaEdgeShape;
(function (OgmaEdgeShape) {
    OgmaEdgeShape["LINE"] = "line";
    OgmaEdgeShape["ARROW"] = "arrow";
    OgmaEdgeShape["TAPERED"] = "tapered";
    OgmaEdgeShape["DASHED"] = "dashed";
    OgmaEdgeShape["DOTTED"] = "dotted";
})(OgmaEdgeShape = exports.OgmaEdgeShape || (exports.OgmaEdgeShape = {}));
