// @linkurious/rest-client/dist/src/api/Application/index.js
"use strict";
/**
 * Copyright Linkurious SAS 2012 - 2019
 *
 * - Created on 2019-10-30.
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var request_1 = require("../../http/request");
var response_1 = require("../../http/response");
__export(require("./types"));
var UNAUTHORIZED = response_1.LkErrorKey.UNAUTHORIZED, FORBIDDEN = response_1.LkErrorKey.FORBIDDEN, NOT_FOUND = response_1.LkErrorKey.NOT_FOUND;
var ApplicationAPI = /** @class */ (function (_super) {
    __extends(ApplicationAPI, _super);
    function ApplicationAPI() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Get all the API applications.
     */
    ApplicationAPI.prototype.getApplications = function () {
        return this.request({
            errors: [UNAUTHORIZED, FORBIDDEN],
            url: '/admin/applications',
            method: 'GET'
        });
    };
    /**
     * Add a new API application.
     */
    ApplicationAPI.prototype.createApplication = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, FORBIDDEN, NOT_FOUND],
            url: '/admin/applications',
            method: 'POST',
            params: params
        });
    };
    /**
     * Update an API application.
     */
    ApplicationAPI.prototype.updateApplication = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, FORBIDDEN, NOT_FOUND],
            url: '/admin/applications/:id',
            method: 'PATCH',
            params: params
        });
    };
    return ApplicationAPI;
}(request_1.Request));
exports.ApplicationAPI = ApplicationAPI;
