// @linkurious/rest-client/dist/src/api/Application/types.js
"use strict";
/**
 * Copyright Linkurious SAS 2012 - 2019
 *
 * - Created on 2019-10-30.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var ApiRight;
(function (ApiRight) {
    // visualization actions
    ApiRight["VISUALIZATION_READ"] = "visualization.read";
    ApiRight["VISUALIZATION_CREATE"] = "visualization.create";
    ApiRight["VISUALIZATION_DUPLICATE"] = "visualization.duplicate";
    ApiRight["VISUALIZATION_EDIT"] = "visualization.edit";
    ApiRight["VISUALIZATION_DELETE"] = "visualization.delete";
    ApiRight["VISUALIZATION_LIST"] = "visualization.list";
    // visualization folder actions
    ApiRight["VISUALIZATION_FOLDER_CREATE"] = "visualizationFolder.create";
    ApiRight["VISUALIZATION_FOLDER_EDIT"] = "visualizationFolder.edit";
    ApiRight["VISUALIZATION_FOLDER_DELETE"] = "visualizationFolder.delete";
    // visualization sharing actions
    ApiRight["VISUALIZATION_SHARE_READ"] = "visualizationShare.read";
    ApiRight["VISUALIZATION_SHARE_CREATE"] = "visualizationShare.create";
    ApiRight["VISUALIZATION_SHARE_DELETE"] = "visualizationShare.delete";
    // sandbox action
    ApiRight["SANDBOX"] = "sandbox";
    // widget actions
    ApiRight["WIDGET_READ"] = "widget.read";
    ApiRight["WIDGET_CREATE"] = "widget.create";
    ApiRight["WIDGET_EDIT"] = "widget.edit";
    ApiRight["WIDGET_DELETE"] = "widget.delete";
    // graph item actions
    ApiRight["GRAPH_ITEM_READ"] = "graphItem.read";
    ApiRight["GRAPH_ITEM_CREATE"] = "graphItem.create";
    ApiRight["GRAPH_ITEM_EDIT"] = "graphItem.edit";
    ApiRight["GRAPH_ITEM_DELETE"] = "graphItem.delete";
    ApiRight["GRAPH_ITEM_SEARCH"] = "graphItem.search";
    // saved graph query actions
    ApiRight["SAVED_GRAPH_QUERY_READ"] = "savedGraphQuery.read";
    ApiRight["SAVED_GRAPH_QUERY_CREATE"] = "savedGraphQuery.create";
    ApiRight["SAVED_GRAPH_QUERY_EDIT"] = "savedGraphQuery.edit";
    ApiRight["SAVED_GRAPH_QUERY_DELETE"] = "savedGraphQuery.delete";
    ApiRight["GRAPH_RUN_QUERY"] = "graph.runQuery";
    // alert actions
    ApiRight["ALERT_READ"] = "alert.read";
    ApiRight["ALERT_DO_ACTION"] = "alert.doAction";
    ApiRight["ADMIN_ALERTS"] = "admin.alerts";
    // schema action
    ApiRight["SCHEMA"] = "schema";
    // index action
    ApiRight["ADMIN_INDEX"] = "admin.index";
})(ApiRight = exports.ApiRight || (exports.ApiRight = {}));
