// @linkurious/rest-client/dist/src/api/GraphSchema/index.js
"use strict";
/**
 * Copyright Linkurious SAS 2012 - 2019
 *
 * - Created on 2019-05-15.
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var request_1 = require("../../http/request");
var response_1 = require("../../http/response");
__export(require("./types"));
var UNAUTHORIZED = response_1.LkErrorKey.UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE = response_1.LkErrorKey.DATA_SOURCE_UNAVAILABLE, GUEST_DISABLED = response_1.LkErrorKey.GUEST_DISABLED, FORBIDDEN = response_1.LkErrorKey.FORBIDDEN, NOT_FOUND = response_1.LkErrorKey.NOT_FOUND, STRICT_SCHEMA_REQUIRED = response_1.LkErrorKey.STRICT_SCHEMA_REQUIRED;
var GraphSchemaAPI = /** @class */ (function (_super) {
    __extends(GraphSchemaAPI, _super);
    function GraphSchemaAPI() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Start the schema sampling.
     */
    GraphSchemaAPI.prototype.startSchemaSampling = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, FORBIDDEN],
            url: '/admin/:sourceKey/schema/sampling/start',
            method: 'POST',
            params: params
        });
    };
    /**
     * Get the schema sampling status.
     */
    GraphSchemaAPI.prototype.getSamplingStatus = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, GUEST_DISABLED],
            url: '/:sourceKey/schema/sampling/status',
            method: 'GET',
            params: params
        });
    };
    /**
     * Stop the schema sampling.
     */
    GraphSchemaAPI.prototype.stopSchemaSampling = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, FORBIDDEN, DATA_SOURCE_UNAVAILABLE],
            url: '/admin/:sourceKey/schema/sampling/stop',
            method: 'POST',
            params: params
        });
    };
    /**
     * List all `edgeTypes`, `nodeCategories`, `edgeProperties`, `nodeProperties`
     * that exist in the graph database.
     */
    GraphSchemaAPI.prototype.getSimpleSchema = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, GUEST_DISABLED],
            url: '/:sourceKey/graph/schema/simple',
            method: 'GET',
            params: params
        });
    };
    /**
     * Update the strict schema settings of the data-source.
     */
    GraphSchemaAPI.prototype.updateSchemaSettings = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, FORBIDDEN, STRICT_SCHEMA_REQUIRED],
            url: '/admin/:sourceKey/graph/schema/settings',
            method: 'PATCH',
            params: params
        });
    };
    /**
     * Get the list of edge properties that are not indexed for the given data-source.
     */
    GraphSchemaAPI.prototype.getNonIndexedEdgeProperties = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, FORBIDDEN, DATA_SOURCE_UNAVAILABLE],
            url: '/admin/source/:sourceKey/noIndex/edgeProperties',
            method: 'GET',
            params: params
        });
    };
    /**
     * Get the list of node properties that are not indexed for the given data-source.
     */
    GraphSchemaAPI.prototype.getNonIndexedNodeProperties = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, FORBIDDEN, DATA_SOURCE_UNAVAILABLE],
            url: '/admin/source/:sourceKey/noIndex/nodeProperties',
            method: 'GET',
            params: params
        });
    };
    /**
     * Set the list of edge properties that are not indexed for the given data-source.
     */
    GraphSchemaAPI.prototype.setNonIndexedEdgeProperties = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, FORBIDDEN, DATA_SOURCE_UNAVAILABLE],
            url: '/admin/source/:sourceKey/noIndex/edgeProperties',
            method: 'PUT',
            params: params
        });
    };
    /**
     * Set the list of node properties that are not indexed for the given data-source.
     */
    GraphSchemaAPI.prototype.setNonIndexedNodeProperties = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, FORBIDDEN, DATA_SOURCE_UNAVAILABLE],
            url: '/admin/source/:sourceKey/noIndex/nodeProperties',
            method: 'PUT',
            params: params
        });
    };
    /**
     * Add a new type to the graph schema.
     */
    GraphSchemaAPI.prototype.createType = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, FORBIDDEN],
            url: '/admin/:sourceKey/graph/schema/:entityType/types',
            method: 'POST',
            params: params
        });
    };
    /**
     * Update an existing graph schema type.
     */
    GraphSchemaAPI.prototype.updateType = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, FORBIDDEN, NOT_FOUND],
            url: '/admin/:sourceKey/graph/schema/:entityType/types',
            method: 'PATCH',
            params: params
        });
    };
    /**
     * Add a new property for a type on the graph schema.
     */
    GraphSchemaAPI.prototype.createProperty = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, FORBIDDEN],
            url: '/admin/:sourceKey/graph/schema/:entityType/properties',
            method: 'POST',
            params: params
        });
    };
    /**
     * Update an existing graph schema property.
     */
    GraphSchemaAPI.prototype.updateProperty = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, FORBIDDEN, NOT_FOUND],
            url: '/admin/:sourceKey/graph/schema/:entityType/properties',
            method: 'PATCH',
            params: params
        });
    };
    /**
     * List all the types and properties of a data-source.
     */
    GraphSchemaAPI.prototype.getTypes = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, FORBIDDEN],
            url: '/admin/:sourceKey/graph/schema/:entityType/types',
            method: 'GET',
            params: params
        });
    };
    /**
     * List all the readable types and properties of a data-source.
     */
    GraphSchemaAPI.prototype.getTypesWithAccess = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, GUEST_DISABLED],
            url: '/:sourceKey/graph/schema/:entityType/types',
            method: 'GET',
            params: params
        });
    };
    return GraphSchemaAPI;
}(request_1.Request));
exports.GraphSchemaAPI = GraphSchemaAPI;
