// @linkurious/rest-client/dist/src/api/GraphSchema/types.js
"use strict";
/**
 * Copyright Linkurious SAS 2012 - 2019
 *
 * - Created on 2019-05-15.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var SamplingStatus;
(function (SamplingStatus) {
    SamplingStatus["ONGOING"] = "ongoing";
    SamplingStatus["DONE"] = "done";
    SamplingStatus["ERROR"] = "error";
})(SamplingStatus = exports.SamplingStatus || (exports.SamplingStatus = {}));
var EntityType;
(function (EntityType) {
    EntityType["NODE"] = "node";
    EntityType["EDGE"] = "edge";
})(EntityType = exports.EntityType || (exports.EntityType = {}));
var DataVisibility;
(function (DataVisibility) {
    DataVisibility["NONE"] = "none";
    DataVisibility["AVAILABLE"] = "available";
    DataVisibility["SEARCHABLE"] = "searchable";
})(DataVisibility = exports.DataVisibility || (exports.DataVisibility = {}));
var PropertyTypeName;
(function (PropertyTypeName) {
    PropertyTypeName["AUTO"] = "auto";
    PropertyTypeName["BOOLEAN"] = "boolean";
    PropertyTypeName["DATE"] = "date";
    PropertyTypeName["DATETIME"] = "datetime";
    PropertyTypeName["NUMBER"] = "number";
    PropertyTypeName["STRING"] = "string";
})(PropertyTypeName = exports.PropertyTypeName || (exports.PropertyTypeName = {}));
var DateFormat;
(function (DateFormat) {
    DateFormat["NATIVE"] = "native";
    DateFormat["ISO"] = "iso";
    DateFormat["DD_MM_YYYY"] = "dd/mm/yyyy";
    DateFormat["MM_DD_YYYY"] = "mm/dd/yyyy";
    DateFormat["TIMESTAMP"] = "timestamp";
    DateFormat["TIMESTAMP_MS"] = "timestamp-ms";
})(DateFormat = exports.DateFormat || (exports.DateFormat = {}));
var DateTimeFormat;
(function (DateTimeFormat) {
    DateTimeFormat["NATIVE"] = "native";
    DateTimeFormat["ISO"] = "iso";
    DateTimeFormat["TIMESTAMP"] = "timestamp";
    DateTimeFormat["TIMESTAMP_MS"] = "timestamp-ms";
})(DateTimeFormat = exports.DateTimeFormat || (exports.DateTimeFormat = {}));
