// @linkurious/rest-client/dist/src/api/Visualization/index.js
"use strict";
/**
 * Copyright Linkurious SAS 2012 - 2019
 *
 * - Created on 2019-10-30.
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var request_1 = require("../../http/request");
var response_1 = require("../../http/response");
__export(require("./types"));
var UNAUTHORIZED = response_1.LkErrorKey.UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE = response_1.LkErrorKey.DATA_SOURCE_UNAVAILABLE, GUEST_DISABLED = response_1.LkErrorKey.GUEST_DISABLED, NOT_FOUND = response_1.LkErrorKey.NOT_FOUND, FOLDER_DELETION_FAILED = response_1.LkErrorKey.FOLDER_DELETION_FAILED, ALREADY_EXISTS = response_1.LkErrorKey.ALREADY_EXISTS, VISUALIZATION_LOCKED = response_1.LkErrorKey.VISUALIZATION_LOCKED;
var VisualizationAPI = /** @class */ (function (_super) {
    __extends(VisualizationAPI, _super);
    function VisualizationAPI() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Get the number of visualizations for the current user in this data-source.
     */
    VisualizationAPI.prototype.getVisualizationCount = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, GUEST_DISABLED],
            url: '/:sourceKey/visualizations/count',
            method: 'GET',
            params: params
        });
    };
    /**
     * Get a visualization by id.
     */
    VisualizationAPI.prototype.getVisualization = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, NOT_FOUND],
            url: '/:sourceKey/visualizations/:id',
            method: 'GET',
            params: params
        });
    };
    /**
     * Create a new visualization.
     */
    VisualizationAPI.prototype.createVisualization = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE],
            url: '/:sourceKey/visualizations',
            method: 'POST',
            params: params
        });
    };
    /**
     * Duplicate a visualization. Return the id of the copy.
     */
    VisualizationAPI.prototype.duplicateVisualization = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, NOT_FOUND],
            url: '/:sourceKey/visualizations/:id/duplicate',
            method: 'POST',
            params: params
        });
    };
    /**
     * Delete the visualization selected by id.
     */
    VisualizationAPI.prototype.deleteVisualization = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, NOT_FOUND],
            url: '/:sourceKey/visualizations/:id',
            method: 'DELETE',
            params: params
        });
    };
    /**
     * Update the visualization selected by id.
     */
    VisualizationAPI.prototype.updateVisualization = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, NOT_FOUND, VISUALIZATION_LOCKED],
            url: '/:sourceKey/visualizations/:id',
            method: 'PATCH',
            params: params
        });
    };
    /**
     * Get all visualizations shared with the current user.
     */
    VisualizationAPI.prototype.getSharedVisualizations = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE],
            url: '/:sourceKey/visualizations/shared',
            method: 'GET',
            params: params
        });
    };
    /**
     * Create a visualization folder.
     */
    VisualizationAPI.prototype.createVisualizationFolder = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, ALREADY_EXISTS],
            url: '/:sourceKey/visualizations/folder',
            method: 'POST',
            params: params
        });
    };
    /**
     * Update a visualization folder.
     */
    VisualizationAPI.prototype.updateVisualizationFolder = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, NOT_FOUND, ALREADY_EXISTS],
            url: '/:sourceKey/visualizations/folder/:id',
            method: 'PATCH',
            params: params
        });
    };
    /**
     * Delete a visualization folder.
     */
    VisualizationAPI.prototype.deleteVisualizationFolder = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, NOT_FOUND, FOLDER_DELETION_FAILED],
            url: '/:sourceKey/visualizations/folder/:id',
            method: 'DELETE',
            params: params
        });
    };
    /**
     * Get the visualizations and the visualization folders in a tree structure.
     */
    VisualizationAPI.prototype.getVisualizationTree = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE],
            url: '/:sourceKey/visualizations/tree',
            method: 'GET',
            params: params
        });
    };
    /**
     * Get the visualization sandbox of the current user for a given data-source.
     */
    VisualizationAPI.prototype.getSandbox = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, GUEST_DISABLED, NOT_FOUND],
            url: '/:sourceKey/sandbox',
            method: 'GET',
            params: params
        });
    };
    /**
     * Update the sandbox of the current user for a given data-source.
     */
    VisualizationAPI.prototype.updateSandbox = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE],
            url: '/:sourceKey/sandbox',
            method: 'PATCH',
            params: params
        });
    };
    /**
     * Get all share rights on a visualization.
     */
    VisualizationAPI.prototype.getVisualizationShares = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, NOT_FOUND],
            url: '/:sourceKey/visualizations/:id/shares',
            method: 'GET',
            params: params
        });
    };
    /**
     * Set the share right of a user on a visualization.
     */
    VisualizationAPI.prototype.shareVisualization = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, NOT_FOUND],
            url: "/:sourceKey/visualizations/:id/share/:userId",
            method: 'PUT',
            params: params
        });
    };
    /**
     * Remove a share right of a user on a visualization.
     */
    VisualizationAPI.prototype.unshareVisualization = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, NOT_FOUND],
            url: "/:sourceKey/visualizations/:id/share/:userId",
            method: 'DELETE',
            params: params
        });
    };
    /**
     * Get a visualization widget's data by key.
     */
    VisualizationAPI.prototype.getWidget = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, NOT_FOUND],
            url: '/widget/:widgetKey',
            method: 'GET',
            params: params
        });
    };
    /**
     * Create a widget for a visualization. Return the widget key.
     */
    VisualizationAPI.prototype.createWidget = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE],
            url: '/widget',
            method: 'POST',
            params: params
        });
    };
    /**
     * Update the widget of a visualization. Return the widget key.
     */
    VisualizationAPI.prototype.updateWidget = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, NOT_FOUND],
            url: '/widget',
            method: 'PUT',
            params: params
        });
    };
    /**
     * Delete the widget of a visualization.
     */
    VisualizationAPI.prototype.deleteWidget = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, NOT_FOUND],
            url: '/widget/:widgetKey',
            method: 'DELETE',
            params: params
        });
    };
    return VisualizationAPI;
}(request_1.Request));
exports.VisualizationAPI = VisualizationAPI;
