// @linkurious/rest-client/dist/src/api/Visualization/types.js
"use strict";
/**
 * Copyright Linkurious SAS 2012 - 2019
 *
 * - Created on 2019-10-30.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var VisualizationRight;
(function (VisualizationRight) {
    VisualizationRight["READ"] = "read";
    VisualizationRight["WRITE"] = "write";
    VisualizationRight["WRITE_FILTERED"] = "write-filtered";
    VisualizationRight["OWNER"] = "owner";
    VisualizationRight["OWNER_FILTERED"] = "owner-filtered";
})(VisualizationRight = exports.VisualizationRight || (exports.VisualizationRight = {}));
var VisualizationMode;
(function (VisualizationMode) {
    VisualizationMode["NODE_LINK"] = "nodelink";
    VisualizationMode["GEO"] = "geo";
})(VisualizationMode = exports.VisualizationMode || (exports.VisualizationMode = {}));
var LayoutAlgorithm;
(function (LayoutAlgorithm) {
    LayoutAlgorithm["FORCE"] = "force";
    LayoutAlgorithm["HIERARCHICAL"] = "hierarchical";
    LayoutAlgorithm["RADIAL"] = "radial";
})(LayoutAlgorithm = exports.LayoutAlgorithm || (exports.LayoutAlgorithm = {}));
var ZoomLevel;
(function (ZoomLevel) {
    ZoomLevel["YEARS"] = "years";
    ZoomLevel["MONTHS"] = "months";
    ZoomLevel["DAYS"] = "days";
    ZoomLevel["HOURS"] = "hours";
    ZoomLevel["MINUTES"] = "minutes";
    ZoomLevel["SECONDS"] = "seconds";
})(ZoomLevel = exports.ZoomLevel || (exports.ZoomLevel = {}));
var PopulateType;
(function (PopulateType) {
    PopulateType["VISUALIZATION_ID"] = "visualizationId";
    PopulateType["EXPAND_NODE_ID"] = "expandNodeId";
    PopulateType["NODE_ID"] = "nodeId";
    PopulateType["EDGE_ID"] = "edgeId";
    PopulateType["SEARCH_NODE"] = "searchNodes";
    PopulateType["SEARCH_EDGE"] = "searchEdges";
    PopulateType["PATTERN"] = "pattern";
    PopulateType["MATCH_ID"] = "matchId";
})(PopulateType = exports.PopulateType || (exports.PopulateType = {}));
