// @linkurious/rest-client/dist/src/api/Alert/index.js
"use strict";
/**
 * Copyright Linkurious SAS 2012 - 2019
 *
 * - Created on 2019-10-30.
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var request_1 = require("../../http/request");
var response_1 = require("../../http/response");
__export(require("./types"));
var FEATURE_DISABLED = response_1.LkErrorKey.FEATURE_DISABLED, UNAUTHORIZED = response_1.LkErrorKey.UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE = response_1.LkErrorKey.DATA_SOURCE_UNAVAILABLE, FORBIDDEN = response_1.LkErrorKey.FORBIDDEN, NOT_FOUND = response_1.LkErrorKey.NOT_FOUND, BAD_GRAPH_REQUEST = response_1.LkErrorKey.BAD_GRAPH_REQUEST, GRAPH_REQUEST_TIMEOUT = response_1.LkErrorKey.GRAPH_REQUEST_TIMEOUT, CONSTRAINT_VIOLATION = response_1.LkErrorKey.CONSTRAINT_VIOLATION, FOLDER_DELETION_FAILED = response_1.LkErrorKey.FOLDER_DELETION_FAILED, ALREADY_EXISTS = response_1.LkErrorKey.ALREADY_EXISTS;
var AlertAPI = /** @class */ (function (_super) {
    __extends(AlertAPI, _super);
    function AlertAPI() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Create a new alert. If `matchTTL` is set to 0, unconfirmed matches
     * will disappear when they stop matching the alert query.
     */
    AlertAPI.prototype.createAlert = function (params) {
        return this.request({
            errors: [FEATURE_DISABLED, UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, FORBIDDEN],
            url: '/admin/:sourceKey/alerts',
            method: 'POST',
            params: params
        });
    };
    /**
     * Update the alert selected by id.
     * Updating an alert query will results in all the previous detected matches deleted.
     */
    AlertAPI.prototype.updateAlert = function (params) {
        return this.request({
            errors: [FEATURE_DISABLED, UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, FORBIDDEN, NOT_FOUND],
            url: '/admin/:sourceKey/alerts/:id',
            method: 'PATCH',
            params: params
        });
    };
    /**
     * Delete the alert by id and all its matches.
     */
    AlertAPI.prototype.deleteAlert = function (params) {
        return this.request({
            errors: [FEATURE_DISABLED, UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, FORBIDDEN, NOT_FOUND],
            url: '/admin/:sourceKey/alerts/:id',
            method: 'DELETE',
            params: params
        });
    };
    /**
     * Create an alert folder.
     */
    AlertAPI.prototype.createAlertFolder = function (params) {
        return this.request({
            errors: [FEATURE_DISABLED, UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, FORBIDDEN, ALREADY_EXISTS],
            url: '/admin/:sourceKey/alerts/folder',
            method: 'POST',
            params: params
        });
    };
    /**
     * Update an alert folder.
     */
    AlertAPI.prototype.updateAlertFolder = function (params) {
        return this.request({
            errors: [
                FEATURE_DISABLED,
                UNAUTHORIZED,
                DATA_SOURCE_UNAVAILABLE,
                FORBIDDEN,
                NOT_FOUND,
                ALREADY_EXISTS
            ],
            url: '/admin/:sourceKey/alerts/folder/:id',
            method: 'PATCH',
            params: params
        });
    };
    /**
     * Delete an alert folder.
     */
    AlertAPI.prototype.deleteAlertFolder = function (params) {
        return this.request({
            errors: [
                FEATURE_DISABLED,
                UNAUTHORIZED,
                DATA_SOURCE_UNAVAILABLE,
                FORBIDDEN,
                NOT_FOUND,
                FOLDER_DELETION_FAILED
            ],
            url: '/admin/:sourceKey/alerts/folder/:id',
            method: 'DELETE',
            params: params
        });
    };
    /**
     * Get the alerts and the alert folders in a tree structure.
     */
    AlertAPI.prototype.getAlertTree = function (params) {
        return this.request({
            errors: [FEATURE_DISABLED, UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, FORBIDDEN],
            url: '/:sourceKey/alerts/tree',
            method: 'GET',
            params: params
        });
    };
    /**
     * Get an alert by id.
     */
    AlertAPI.prototype.getAlert = function (params) {
        return this.request({
            errors: [FEATURE_DISABLED, UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, FORBIDDEN, NOT_FOUND],
            url: '/:sourceKey/alerts/:id',
            method: 'GET',
            params: params
        });
    };
    /**
     * Get a match by id.
     */
    AlertAPI.prototype.getMatch = function (params) {
        return this.request({
            errors: [FEATURE_DISABLED, UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, FORBIDDEN, NOT_FOUND],
            url: '/:sourceKey/alerts/:alertId/matches/:matchId',
            method: 'GET',
            params: params
        });
    };
    /**
     * Get all the matches of an alert.
     */
    AlertAPI.prototype.getMatches = function (params) {
        return this.request({
            errors: [FEATURE_DISABLED, UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, FORBIDDEN, NOT_FOUND],
            url: '/:sourceKey/alerts/:alertId/matches',
            method: 'GET',
            params: params
        });
    };
    /**
     * Get all the actions of a match ordered by creation date.
     */
    AlertAPI.prototype.getMatchActions = function (params) {
        return this.request({
            errors: [FEATURE_DISABLED, UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, FORBIDDEN, NOT_FOUND],
            url: '/:sourceKey/alerts/:alertId/matches/:matchId/actions',
            method: 'GET',
            params: params
        });
    };
    /**
     * Do an action (open, dismiss, confirm, unconfirm) on a match.
     */
    AlertAPI.prototype.doMatchAction = function (params) {
        return this.request({
            errors: [FEATURE_DISABLED, UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, FORBIDDEN, NOT_FOUND],
            url: '/:sourceKey/alerts/:alertId/matches/:matchId/action',
            method: 'POST',
            params: params
        });
    };
    /**
     * Get all the nodes and edges matching the given graph query.
     * An array of subgraphs, one for each subgraph matching the graph query, is returned.
     */
    AlertAPI.prototype.alertPreview = function (params) {
        return this.request({
            errors: [
                FEATURE_DISABLED,
                UNAUTHORIZED,
                DATA_SOURCE_UNAVAILABLE,
                FORBIDDEN,
                BAD_GRAPH_REQUEST,
                GRAPH_REQUEST_TIMEOUT,
                CONSTRAINT_VIOLATION
            ],
            url: '/:sourceKey/graph/alertPreview',
            method: 'POST',
            params: params
        });
    };
    return AlertAPI;
}(request_1.Request));
exports.AlertAPI = AlertAPI;
