// @linkurious/rest-client/dist/src/api/Alert/types.js
"use strict";
/**
 * Copyright Linkurious SAS 2012 - 2019
 *
 * - Created on 2019-10-30.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var AlertColumnType;
(function (AlertColumnType) {
    AlertColumnType["STRING"] = "string";
    AlertColumnType["NUMBER"] = "number";
})(AlertColumnType = exports.AlertColumnType || (exports.AlertColumnType = {}));
var MatchStatus;
(function (MatchStatus) {
    MatchStatus["UNCONFIRMED"] = "unconfirmed";
    MatchStatus["CONFIRMED"] = "confirmed";
    MatchStatus["DISMISSED"] = "dismissed";
})(MatchStatus = exports.MatchStatus || (exports.MatchStatus = {}));
var GetMatchesSortDirection;
(function (GetMatchesSortDirection) {
    GetMatchesSortDirection["ASC"] = "asc";
    GetMatchesSortDirection["DESC"] = "desc";
})(GetMatchesSortDirection = exports.GetMatchesSortDirection || (exports.GetMatchesSortDirection = {}));
var GetMatchesSortBy;
(function (GetMatchesSortBy) {
    GetMatchesSortBy["DATE"] = "date";
    GetMatchesSortBy["ZERO"] = "0";
    GetMatchesSortBy["ONE"] = "1";
    GetMatchesSortBy["TWO"] = "2";
    GetMatchesSortBy["THREE"] = "3";
    GetMatchesSortBy["FOUR"] = "4";
})(GetMatchesSortBy = exports.GetMatchesSortBy || (exports.GetMatchesSortBy = {}));
var MatchActionType;
(function (MatchActionType) {
    MatchActionType["CONFIRM"] = "confirm";
    MatchActionType["DISMISS"] = "dismiss";
    MatchActionType["UNCONFIRM"] = "unconfirm";
    MatchActionType["OPEN"] = "open";
})(MatchActionType = exports.MatchActionType || (exports.MatchActionType = {}));
