// @linkurious/rest-client/dist/src/api/AccessRight/types.js
"use strict";
/**
 * Copyright Linkurious SAS 2012 - 2019
 *
 * - Created on 2019-09-27.
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Available to custom groups.
 */
var Action;
(function (Action) {
    Action["ADMIN_CONNECT"] = "admin.connect";
    Action["ADMIN_INDEX"] = "admin.index";
    Action["ADMIN_USERS"] = "admin.users";
    Action["ADMIN_SCHEMA"] = "admin.schema";
    Action["ADMIN_ALERTS"] = "admin.alerts";
    Action["ADMIN_REPORT"] = "admin.report";
    Action["ADMIN_STYLES"] = "admin.styles";
    Action["RUN_QUERY"] = "runQuery";
    Action["RAW_READ_QUERY"] = "rawReadQuery";
    Action["RAW_WRITE_QUERY"] = "rawWriteQuery";
    Action["RUN_CUSTOM_ACTION"] = "runCustomAction";
    Action["WRITE_CUSTOM_ACTION"] = "writeCustomAction"; // Edit, update and delete a custom action
})(Action = exports.Action || (exports.Action = {}));
/**
 * Only for the built-in admin group.
 */
var AdminAction;
(function (AdminAction) {
    AdminAction["ADMIN_APP"] = "admin.app";
    AdminAction["ADMIN_USERS_DELETE"] = "admin.users.delete";
    AdminAction["ADMIN_CONFIG"] = "admin.config"; // Edit the configuration of Linkurious
})(AdminAction = exports.AdminAction || (exports.AdminAction = {}));
var ItemTypeAccessRightType;
(function (ItemTypeAccessRightType) {
    ItemTypeAccessRightType["READ"] = "read";
    ItemTypeAccessRightType["EDIT"] = "edit";
    ItemTypeAccessRightType["WRITE"] = "write";
    ItemTypeAccessRightType["NONE"] = "none";
})(ItemTypeAccessRightType = exports.ItemTypeAccessRightType || (exports.ItemTypeAccessRightType = {}));
var PropertyAccessRightType;
(function (PropertyAccessRightType) {
    PropertyAccessRightType["READ"] = "read";
    PropertyAccessRightType["EDIT"] = "edit";
    PropertyAccessRightType["NONE"] = "none";
})(PropertyAccessRightType = exports.PropertyAccessRightType || (exports.PropertyAccessRightType = {}));
var ActionAccessRightType;
(function (ActionAccessRightType) {
    ActionAccessRightType["DO"] = "do";
    ActionAccessRightType["NONE"] = "none";
})(ActionAccessRightType = exports.ActionAccessRightType || (exports.ActionAccessRightType = {}));
var AlertAccessRightType;
(function (AlertAccessRightType) {
    AlertAccessRightType["READ"] = "read";
    AlertAccessRightType["NONE"] = "none";
})(AlertAccessRightType = exports.AlertAccessRightType || (exports.AlertAccessRightType = {}));
var TargetType;
(function (TargetType) {
    TargetType["NODE_CATEGORY"] = "nodeCategory";
    TargetType["EDGE_TYPE"] = "edgeType";
    TargetType["NODE_PROPERTY_KEY"] = "nodePropertyKey";
    TargetType["EDGE_PROPERTY_KEY"] = "edgePropertyKey";
    TargetType["ACTION"] = "action";
    TargetType["ALERT"] = "alert";
})(TargetType = exports.TargetType || (exports.TargetType = {}));
