// @linkurious/rest-client/dist/src/api/GraphNode/index.js
"use strict";
/**
 * Copyright Linkurious SAS 2012 - 2019
 *
 * - Created on 2019-10-30.
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var request_1 = require("../../http/request");
var response_1 = require("../../http/response");
__export(require("./types"));
var UNAUTHORIZED = response_1.LkErrorKey.UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE = response_1.LkErrorKey.DATA_SOURCE_UNAVAILABLE, GUEST_DISABLED = response_1.LkErrorKey.GUEST_DISABLED, FORBIDDEN = response_1.LkErrorKey.FORBIDDEN, NOT_FOUND = response_1.LkErrorKey.NOT_FOUND, EDIT_CONFLICT = response_1.LkErrorKey.EDIT_CONFLICT, NOT_SUPPORTED = response_1.LkErrorKey.NOT_SUPPORTED;
var GraphNodeAPI = /** @class */ (function (_super) {
    __extends(GraphNodeAPI, _super);
    function GraphNodeAPI() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Get a node of the graph.
     * A subgraph made of the single node is returned.
     */
    GraphNodeAPI.prototype.getNode = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, GUEST_DISABLED, NOT_FOUND],
            url: '/:sourceKey/graph/nodes/:id',
            method: 'POST',
            params: params
        });
    };
    /**
     * Add a node to the graph.
     */
    GraphNodeAPI.prototype.createNode = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, FORBIDDEN],
            url: '/:sourceKey/graph/nodes',
            method: 'POST',
            params: params
        });
    };
    /**
     * Update a subset of properties and categories of a node.
     * Keep every other property and category of the node unchanged.
     */
    GraphNodeAPI.prototype.updateNode = function (params) {
        return this.request({
            errors: [
                UNAUTHORIZED,
                DATA_SOURCE_UNAVAILABLE,
                FORBIDDEN,
                NOT_FOUND,
                EDIT_CONFLICT,
                NOT_SUPPORTED
            ],
            url: '/:sourceKey/graph/nodes/:id',
            method: 'PATCH',
            params: params
        });
    };
    /**
     * Delete a node and its adjacent edges from the graph.
     */
    GraphNodeAPI.prototype.deleteNode = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, FORBIDDEN, NOT_FOUND],
            url: '/:sourceKey/graph/nodes/:id',
            method: 'DELETE',
            params: params
        });
    };
    /**
     * Get the number of nodes in the graph.
     */
    GraphNodeAPI.prototype.getNodeCount = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, GUEST_DISABLED],
            url: '/:sourceKey/graph/nodes/count',
            method: 'GET',
            params: params
        });
    };
    /**
     * Get the digest (the number of adjacent nodes and edges grouped by node categories and edge types)
     * and/or the degree of a given subset of nodes.
     * You can't get aggregated statistics of a subset of nodes containing one or more supernodes.
     * To get the statistics of a supernode invoke the API with only its node ID.
     */
    GraphNodeAPI.prototype.getStatistics = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, GUEST_DISABLED, NOT_FOUND],
            url: '/:sourceKey/graph/neighborhood/statistics',
            method: 'POST',
            params: params
        });
    };
    /**
     * Get all the adjacent nodes and edges to one or more source nodes.
     * A subgraph made of the items that matched the expand query and the edges between them is returned.
     */
    GraphNodeAPI.prototype.getAdjacentNodes = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, GUEST_DISABLED, NOT_FOUND],
            url: '/:sourceKey/graph/nodes/expand',
            method: 'POST',
            params: params
        });
    };
    return GraphNodeAPI;
}(request_1.Request));
exports.GraphNodeAPI = GraphNodeAPI;
