// @linkurious/rest-client/dist/src/api/GraphQuery/index.js
"use strict";
/**
 * Copyright Linkurious SAS 2012 - 2019
 *
 * - Created on 2019-10-29.
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var request_1 = require("../../http/request");
var response_1 = require("../../http/response");
__export(require("./types"));
var UNAUTHORIZED = response_1.LkErrorKey.UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE = response_1.LkErrorKey.DATA_SOURCE_UNAVAILABLE, GUEST_DISABLED = response_1.LkErrorKey.GUEST_DISABLED, FORBIDDEN = response_1.LkErrorKey.FORBIDDEN, NOT_FOUND = response_1.LkErrorKey.NOT_FOUND, BAD_GRAPH_REQUEST = response_1.LkErrorKey.BAD_GRAPH_REQUEST, GRAPH_REQUEST_TIMEOUT = response_1.LkErrorKey.GRAPH_REQUEST_TIMEOUT, CONSTRAINT_VIOLATION = response_1.LkErrorKey.CONSTRAINT_VIOLATION, MALFORMED_QUERY_TEMPLATE = response_1.LkErrorKey.MALFORMED_QUERY_TEMPLATE;
var GraphQueryAPI = /** @class */ (function (_super) {
    __extends(GraphQueryAPI, _super);
    function GraphQueryAPI() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Get a graph query owned by the current user or shared with it.
     */
    GraphQueryAPI.prototype.getQuery = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, GUEST_DISABLED, FORBIDDEN, NOT_FOUND],
            url: '/:sourceKey/graph/query/:id',
            method: 'GET',
            params: params
        });
    };
    /**
     * Get all the graph queries owned by the current user or shared with it.
     */
    GraphQueryAPI.prototype.getQueries = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, GUEST_DISABLED, FORBIDDEN],
            url: '/:sourceKey/graph/query',
            method: 'GET',
            params: params
        });
    };
    /**
     * Create a graph query for the current user.
     */
    GraphQueryAPI.prototype.createQuery = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, FORBIDDEN, MALFORMED_QUERY_TEMPLATE],
            url: '/:sourceKey/graph/query',
            method: 'POST',
            params: params
        });
    };
    /**
     * Update a graph query owned by the current user.
     */
    GraphQueryAPI.prototype.updateQuery = function (params) {
        return this.request({
            errors: [
                UNAUTHORIZED,
                DATA_SOURCE_UNAVAILABLE,
                FORBIDDEN,
                NOT_FOUND,
                MALFORMED_QUERY_TEMPLATE
            ],
            url: '/:sourceKey/graph/query/:id',
            method: 'PATCH',
            params: params
        });
    };
    /**
     * Delete a graph query owned by the current user.
     */
    GraphQueryAPI.prototype.deleteQuery = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, FORBIDDEN, NOT_FOUND],
            url: '/:sourceKey/graph/query/:id',
            method: 'DELETE',
            params: params
        });
    };
    /**
     * Check that the given graph query is syntactically correct. Parse the query if it's a template.
     */
    GraphQueryAPI.prototype.checkQuery = function (params) {
        return this.request({
            errors: [
                UNAUTHORIZED,
                DATA_SOURCE_UNAVAILABLE,
                FORBIDDEN,
                BAD_GRAPH_REQUEST,
                GRAPH_REQUEST_TIMEOUT,
                CONSTRAINT_VIOLATION,
                MALFORMED_QUERY_TEMPLATE
            ],
            url: '/:sourceKey/graph/check/query',
            method: 'POST',
            params: params
        });
    };
    /**
     * Get all the nodes and edges matching the given graph query.
     * A subgraph made of all the nodes and the edges from each subgraph matching the graph query is returned.
     */
    GraphQueryAPI.prototype.runQuery = function (params) {
        return this.request({
            errors: [
                UNAUTHORIZED,
                DATA_SOURCE_UNAVAILABLE,
                FORBIDDEN,
                NOT_FOUND,
                BAD_GRAPH_REQUEST,
                GRAPH_REQUEST_TIMEOUT,
                CONSTRAINT_VIOLATION,
                MALFORMED_QUERY_TEMPLATE
            ],
            url: '/:sourceKey/graph/run/query',
            method: 'POST',
            params: params
        });
    };
    /**
     * Get all the nodes and edges matching the given saved graph query by id.
     * A subgraph made of all the nodes and the edges from each subgraph matching the graph query is returned.
     */
    GraphQueryAPI.prototype.runQueryById = function (params) {
        return this.request({
            errors: [
                UNAUTHORIZED,
                DATA_SOURCE_UNAVAILABLE,
                GUEST_DISABLED,
                FORBIDDEN,
                NOT_FOUND,
                BAD_GRAPH_REQUEST,
                GRAPH_REQUEST_TIMEOUT,
                CONSTRAINT_VIOLATION
            ],
            url: '/:sourceKey/graph/run/query/:id',
            method: 'POST',
            params: params
        });
    };
    return GraphQueryAPI;
}(request_1.Request));
exports.GraphQueryAPI = GraphQueryAPI;
