// @linkurious/rest-client/dist/src/api/GraphQuery/types.js
"use strict";
/**
 * Copyright Linkurious SAS 2012 - 2019
 *
 * - Created on 2019-10-29.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var TemplateFieldType;
(function (TemplateFieldType) {
    TemplateFieldType["NUMBER"] = "number";
    TemplateFieldType["STRING"] = "string";
    TemplateFieldType["ENUM"] = "enum";
    TemplateFieldType["NODE"] = "node";
    TemplateFieldType["NODE_SET"] = "nodeset";
    TemplateFieldType["DATE"] = "date";
    TemplateFieldType["DATE_TIME"] = "datetime";
    TemplateFieldType["BOOLEAN"] = "boolean";
    TemplateFieldType["ENV"] = "env";
})(TemplateFieldType = exports.TemplateFieldType || (exports.TemplateFieldType = {}));
var DateTemplateFormat;
(function (DateTemplateFormat) {
    DateTemplateFormat["TIMESTAMP"] = "timestamp";
    DateTemplateFormat["TIMESTAMP_MS"] = "timestamp-ms";
    DateTemplateFormat["ISO"] = "iso";
    DateTemplateFormat["ISO_YYYY_MM_DD"] = "yyyy-mm-dd";
    DateTemplateFormat["DD_MM_YYYY"] = "dd/mm/yyyy";
    DateTemplateFormat["MM_DD_YYYY"] = "mm/dd/yyyy";
    DateTemplateFormat["NATIVE"] = "native";
})(DateTemplateFormat = exports.DateTemplateFormat || (exports.DateTemplateFormat = {}));
var EnvTemplateValues;
(function (EnvTemplateValues) {
    EnvTemplateValues["EMAIL"] = "email";
})(EnvTemplateValues = exports.EnvTemplateValues || (exports.EnvTemplateValues = {}));
var DatetimeTemplateFormat;
(function (DatetimeTemplateFormat) {
    DatetimeTemplateFormat["TIMESTAMP"] = "timestamp";
    DatetimeTemplateFormat["TIMESTAMP_MS"] = "timestamp-ms";
    DatetimeTemplateFormat["ISO"] = "iso";
    DatetimeTemplateFormat["YYYY_MM_DD_T"] = "YYYY-MM-DDThh:mm:ss";
    DatetimeTemplateFormat["NATIVE"] = "native";
})(DatetimeTemplateFormat = exports.DatetimeTemplateFormat || (exports.DatetimeTemplateFormat = {}));
var GraphQueryInputType;
(function (GraphQueryInputType) {
    GraphQueryInputType["NONE"] = "none";
    GraphQueryInputType["_1_NODE"] = "1-node";
    GraphQueryInputType["_2_NODES"] = "2-nodes";
    GraphQueryInputType["NODESET"] = "nodeset";
})(GraphQueryInputType = exports.GraphQueryInputType || (exports.GraphQueryInputType = {}));
var GraphQueryDialect;
(function (GraphQueryDialect) {
    GraphQueryDialect["CYPHER"] = "cypher";
    GraphQueryDialect["GREMLIN"] = "gremlin";
})(GraphQueryDialect = exports.GraphQueryDialect || (exports.GraphQueryDialect = {}));
var GraphQuerySharingMode;
(function (GraphQuerySharingMode) {
    GraphQuerySharingMode["PRIVATE"] = "private";
    GraphQuerySharingMode["SOURCE"] = "source";
    GraphQuerySharingMode["GROUPS"] = "groups";
})(GraphQuerySharingMode = exports.GraphQuerySharingMode || (exports.GraphQuerySharingMode = {}));
var GraphQueryRight;
(function (GraphQueryRight) {
    GraphQueryRight["OWNER"] = "owner";
    GraphQueryRight["READ"] = "read";
})(GraphQueryRight = exports.GraphQueryRight || (exports.GraphQueryRight = {}));
var GraphQueryType;
(function (GraphQueryType) {
    GraphQueryType["STATIC"] = "static";
    GraphQueryType["TEMPLATE"] = "template";
})(GraphQueryType = exports.GraphQueryType || (exports.GraphQueryType = {}));
