// @linkurious/rest-client/dist/test/restClient.spec.js
"use strict";
/**
 * Copyright Linkurious SAS 2012 - 2019
 *
 * - Created on 2019-12-05.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var mocha_1 = require("mocha");
var chai_1 = require("chai");
var src_1 = require("../src");
var request_1 = require("../src/http/request");
mocha_1.describe('Rest Client', function () {
    mocha_1.it('Should find the correct data-source by index', function () {
        var sources = [
            { configIndex: 0, connected: false },
            { configIndex: 1, connected: true },
            { configIndex: 3, connected: true }
        ];
        chai_1.assert.deepEqual(
        //@ts-ignore partial interface for sources
        src_1.RestClient.getCurrentSource(sources, { configIndex: 0 }), sources[0]);
        chai_1.assert.deepEqual(
        //@ts-ignore partial interface for sources
        src_1.RestClient.getCurrentSource(sources, { configIndex: 1 }), sources[1]);
        chai_1.assert.deepEqual(
        //@ts-ignore partial interface for sources
        src_1.RestClient.getCurrentSource(sources, { configIndex: 3 }), sources[2]);
        chai_1.assert.deepEqual(
        //@ts-ignore partial interface for sources
        src_1.RestClient.getCurrentSource(sources, { configIndex: 5 }), sources[1]);
    });
    mocha_1.it('Should find the correct data-source by user id', function () {
        var sourceByUserId = new Map([
            ['lk-lastSeenSourceKey-1', 's1'],
            ['lk-lastSeenSourceKey-2', 's2'],
            ['lk-lastSeenSourceKey-3', 's3']
        ]);
        var storage = {
            getItem: function (key) { return sourceByUserId.get(key) || null; }
        };
        var sources = [
            { key: 's1', connected: false },
            { key: 's2', connected: true },
            { key: 's3', connected: true }
        ];
        chai_1.assert.deepEqual(
        //@ts-ignore partial interface for sources
        src_1.RestClient.getCurrentSource(sources, { userId: 1 }, storage), sources[1]);
        chai_1.assert.deepEqual(
        //@ts-ignore partial interface for sources
        src_1.RestClient.getCurrentSource(sources, { userId: 2 }, storage), sources[1]);
        chai_1.assert.deepEqual(
        //@ts-ignore partial interface for sources
        src_1.RestClient.getCurrentSource(sources, { userId: 3 }, storage), sources[2]);
        chai_1.assert.deepEqual(
        //@ts-ignore partial interface for sources
        src_1.RestClient.getCurrentSource(sources, { userId: 5 }, storage), sources[1]);
    });
    mocha_1.it('Should find the correct data-source by sourceKey', function () {
        var sources = [
            { key: 's1', connected: false },
            { key: 's2', connected: true },
            { key: 's3', connected: true }
        ];
        chai_1.assert.deepEqual(
        //@ts-ignore partial interface for sources
        src_1.RestClient.getCurrentSource(sources, { sourceKey: 's1' }), sources[0]);
        chai_1.assert.deepEqual(
        //@ts-ignore partial interface for sources
        src_1.RestClient.getCurrentSource(sources, { sourceKey: 's2' }), sources[1]);
        chai_1.assert.deepEqual(
        //@ts-ignore partial interface for sources
        src_1.RestClient.getCurrentSource(sources, { sourceKey: 's3' }), sources[2]);
        chai_1.assert.deepEqual(
        //@ts-ignore partial interface for sources
        src_1.RestClient.getCurrentSource(sources, { sourceKey: 's5' }), sources[1]);
    });
    mocha_1.it('Should render URL correctly', function () {
        chai_1.assert.deepEqual(
        // @ts-ignore private method and incomplete params
        request_1.Request.renderURL({
            url: '/visualizations/:visualizationId/share/:userId',
            params: {
                visualizationId: 1,
                userId: 2
            }
        }).url, '/visualizations/1/share/2');
    });
});
