// @linkurious/rest-client/dist/src/api/Auth/index.js
"use strict";
/**
 * Copyright Linkurious SAS 2012 - 2019
 *
 * - Created on 2019-08-19.
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var response_1 = require("../../http/response");
var request_1 = require("../../http/request");
var UNAUTHORIZED = response_1.LkErrorKey.UNAUTHORIZED, GUEST_DISABLED = response_1.LkErrorKey.GUEST_DISABLED, FORBIDDEN = response_1.LkErrorKey.FORBIDDEN, ALREADY_EXISTS = response_1.LkErrorKey.ALREADY_EXISTS;
var AuthAPI = /** @class */ (function (_super) {
    __extends(AuthAPI, _super);
    function AuthAPI() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Log a user in by e-mail or username and password and return it.
     */
    AuthAPI.prototype.login = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        delete this.props.clientState.currentSource;
                        delete this.props.clientState.guestMode;
                        delete this.props.clientState.sources;
                        delete this.props.clientState.user;
                        return [4 /*yield*/, this.request({
                                errors: [UNAUTHORIZED],
                                url: '/auth/login',
                                method: 'POST',
                                params: params
                            })];
                    case 1:
                        response = _a.sent();
                        if (response.isSuccess()) {
                            this.props.clientState.user = response.body;
                        }
                        return [2 /*return*/, response];
                }
            });
        });
    };
    /**
     * Log a user in via OAuth2.
     */
    AuthAPI.prototype.loginOAuth2 = function (params) {
        return this.request({
            url: '/auth/sso/return',
            method: 'GET',
            params: params
        });
    };
    /**
     * Log the current user out.
     */
    AuthAPI.prototype.logout = function () {
        return __awaiter(this, void 0, void 0, function () {
            var response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.request({
                            errors: [UNAUTHORIZED],
                            url: '/auth/logout',
                            method: 'GET'
                        })];
                    case 1:
                        response = _a.sent();
                        delete this.props.clientState.currentSource;
                        delete this.props.clientState.guestMode;
                        delete this.props.clientState.sources;
                        delete this.props.clientState.user;
                        return [2 /*return*/, response];
                }
            });
        });
    };
    /**
     * Get the profile of the current user.
     */
    AuthAPI.prototype.getCurrentUser = function () {
        return __awaiter(this, void 0, void 0, function () {
            var response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.request({
                            errors: [UNAUTHORIZED, GUEST_DISABLED],
                            url: '/auth/me',
                            method: 'GET'
                        })];
                    case 1:
                        response = _a.sent();
                        if (response.isSuccess()) {
                            this.props.clientState.user = response.body;
                        }
                        return [2 /*return*/, response];
                }
            });
        });
    };
    /**
     * Update the current user.
     */
    AuthAPI.prototype.updateCurrentUser = function (params) {
        return __awaiter(this, void 0, void 0, function () {
            var response;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.request({
                            errors: [UNAUTHORIZED, FORBIDDEN, ALREADY_EXISTS],
                            url: '/auth/me',
                            method: 'PATCH',
                            params: params
                        })];
                    case 1:
                        response = _a.sent();
                        if (response.isSuccess()) {
                            this.props.clientState.user = response.body;
                        }
                        return [2 /*return*/, response];
                }
            });
        });
    };
    return AuthAPI;
}(request_1.Request));
exports.AuthAPI = AuthAPI;
