// @linkurious/rest-client/dist/src/api/Linkurious/index.js
"use strict";
/**
 * Copyright Linkurious SAS 2012 - 2019
 *
 * - Created on 2019-08-19.
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var response_1 = require("../../http/response");
var request_1 = require("../../http/request");
var UNAUTHORIZED = response_1.LkErrorKey.UNAUTHORIZED, FORBIDDEN = response_1.LkErrorKey.FORBIDDEN;
var LinkuriousAPI = /** @class */ (function (_super) {
    __extends(LinkuriousAPI, _super);
    function LinkuriousAPI() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Get the status of the Linkurious server.
     */
    LinkuriousAPI.prototype.getStatus = function () {
        return this.request({
            url: '/status',
            method: 'GET'
        });
    };
    /**
     * Get Linkurious current version information.
     */
    LinkuriousAPI.prototype.getVersion = function () {
        return this.request({
            url: '/version',
            method: 'GET'
        });
    };
    /**
     * Save an event to the analytics log file. All events follow the Segment Spec.
     */
    LinkuriousAPI.prototype.sendAnalytics = function (params) {
        return this.request({
            url: '/analytics',
            method: 'POST',
            params: params
        });
    };
    /**
     * Collect all the analytics and log files in a compressed tarball and return it.
     */
    LinkuriousAPI.prototype.getReport = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, FORBIDDEN],
            url: '/admin/report',
            method: 'GET',
            params: params
        });
    };
    /**
     * Restart Linkurious. Return the URL of Linkurious after the restart.
     */
    LinkuriousAPI.prototype.restartLinkurious = function () {
        return this.request({ errors: [UNAUTHORIZED, FORBIDDEN], url: '/admin/restart', method: 'POST' });
    };
    /**
     * List all custom files in the specified root directory.
     */
    LinkuriousAPI.prototype.getCustomFiles = function (params) {
        return this.request({
            errors: [UNAUTHORIZED],
            url: '/customFiles',
            method: 'GET',
            params: params
        });
    };
    return LinkuriousAPI;
}(request_1.Request));
exports.LinkuriousAPI = LinkuriousAPI;
