// @linkurious/rest-client/dist/src/errorListener.js
"use strict";
/**
 * Copyright Linkurious SAS 2012 - 2019
 *
 * - Created on 2019-03-13.
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var UnexpectedServerError = /** @class */ (function (_super) {
    __extends(UnexpectedServerError, _super);
    function UnexpectedServerError(originalResponse) {
        var _this = _super.call(this, "Unexpected error: " + JSON.stringify(originalResponse.body)) || this;
        _this.originalResponse = originalResponse;
        return _this;
    }
    return UnexpectedServerError;
}(Error));
exports.UnexpectedServerError = UnexpectedServerError;
var ErrorListener = /** @class */ (function () {
    function ErrorListener() {
        this.listeners = {};
    }
    ErrorListener.prototype.dispatchError = function (key, payload) {
        if (Object.prototype.hasOwnProperty.call(this.listeners, key) &&
            typeof this.listeners[key] === 'function') {
            this.listeners[key](payload);
        }
    };
    ErrorListener.prototype.setErrorListener = function (key, callback) {
        this.listeners[key] = callback;
    };
    ErrorListener.prototype.removeErrorListener = function (key) {
        delete this.listeners[key];
    };
    ErrorListener.prototype.getErrorKeysListened = function () {
        return Object.keys(this.listeners);
    };
    return ErrorListener;
}());
exports.ErrorListener = ErrorListener;
