// @linkurious/rest-client/dist/src/index.js
"use strict";
/**
 * Copyright Linkurious SAS 2012 - 2019
 *
 * - Created on 2019-10-25.
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var request = require("superagent");
var errorListener_1 = require("./errorListener");
var AccessRight_1 = require("./api/AccessRight");
var Alert_1 = require("./api/Alert");
var Application_1 = require("./api/Application");
var Auth_1 = require("./api/Auth");
var Config_1 = require("./api/Config");
var CustomAction_1 = require("./api/CustomAction");
var DataSource_1 = require("./api/DataSource");
var GraphEdge_1 = require("./api/GraphEdge");
var GraphNode_1 = require("./api/GraphNode");
var GraphQuery_1 = require("./api/GraphQuery");
var GraphSchema_1 = require("./api/GraphSchema");
var Linkurious_1 = require("./api/Linkurious");
var Plugin_1 = require("./api/Plugin");
var Search_1 = require("./api/Search");
var User_1 = require("./api/User");
var Visualization_1 = require("./api/Visualization");
var utils_1 = require("./utils");
var RestClient = /** @class */ (function (_super) {
    __extends(RestClient, _super);
    function RestClient(options) {
        var _this = _super.call(this) || this;
        _this.clientState = {};
        _this.moduleProps = {
            baseUrl: options
                ? options.baseUrl && utils_1.endsWith(options.baseUrl, '/')
                    ? options.baseUrl + 'api'
                    : options.baseUrl + '/api'
                : '/api',
            agent: (options && options.agent) || request.agent(),
            clientState: _this.clientState,
            dispatchError: function (key, payload) {
                return _this.dispatchError(key, payload);
            }
        };
        _this.accessRight = new AccessRight_1.AccessRightAPI(_this.moduleProps);
        _this.alert = new Alert_1.AlertAPI(_this.moduleProps);
        _this.application = new Application_1.ApplicationAPI(_this.moduleProps);
        _this.auth = new Auth_1.AuthAPI(_this.moduleProps);
        _this.config = new Config_1.ConfigAPI(_this.moduleProps);
        _this.customAction = new CustomAction_1.CustomActionAPI(_this.moduleProps);
        _this.dataSource = new DataSource_1.DataSourceAPI(_this.moduleProps);
        _this.graphEdge = new GraphEdge_1.GraphEdgeAPI(_this.moduleProps);
        _this.graphNode = new GraphNode_1.GraphNodeAPI(_this.moduleProps);
        _this.graphQuery = new GraphQuery_1.GraphQueryAPI(_this.moduleProps);
        _this.graphSchema = new GraphSchema_1.GraphSchemaAPI(_this.moduleProps);
        _this.linkurious = new Linkurious_1.LinkuriousAPI(_this.moduleProps);
        _this.plugin = new Plugin_1.PluginAPI(_this.moduleProps);
        _this.search = new Search_1.SearchAPI(_this.moduleProps);
        _this.user = new User_1.UserAPI(_this.moduleProps);
        _this.visualization = new Visualization_1.VisualizationAPI(_this.moduleProps);
        return _this;
    }
    /**
     * Login a user and populate the client state with the list of the data-sources.
     */
    RestClient.prototype.init = function (data) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.auth.login(data)];
                    case 1:
                        _a.sent();
                        return [4 /*yield*/, this.dataSource.getDataSources({
                                withCaptions: true,
                                withStyles: true
                            })];
                    case 2:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    RestClient.prototype.setGuestMode = function (guestMode) {
        this.clientState.guestMode = guestMode;
    };
    RestClient.prototype.setCurrentSource = function (dataSource) {
        this.clientState.currentSource = dataSource;
        try {
            if (dataSource.key && this.clientState.user) {
                localStorage.setItem('lk-lastSeenSourceKey-' + this.clientState.user.id, dataSource.key);
            }
        }
        catch (_) {
            // Silently fail if localStorage is not supported
        }
    };
    RestClient.getCurrentSource = function (dataSources, by, storage) {
        if (dataSources.length === 0) {
            throw new Error('RestClient::getCurrentSource - dataSources cannot be empty.');
        }
        if (by) {
            var source = void 0;
            if ('userId' in by) {
                // Return the last seen data-source by the current user if the data-source is connected
                try {
                    var sourceKey_1 = (storage || localStorage).getItem('lk-lastSeenSourceKey-' + by.userId);
                    source = utils_1.find(dataSources, function (s) { return s.connected && s.key === sourceKey_1; });
                }
                catch (_) {
                    source = undefined;
                }
            }
            else if ('sourceKey' in by) {
                // Return the data-source whose sourceKey matches sourceKey in input
                source = utils_1.find(dataSources, function (s) { return s.key === by.sourceKey; });
            }
            else {
                // Return the data-source whose configIndex matches configIndex in input
                source = utils_1.find(dataSources, function (s) { return s.configIndex === by.configIndex; });
            }
            if (source) {
                return source;
            }
        }
        // Return the first connected data-source
        for (var _i = 0, dataSources_1 = dataSources; _i < dataSources_1.length; _i++) {
            var firstConnected = dataSources_1[_i];
            if (firstConnected.connected) {
                return firstConnected;
            }
        }
        // Return the first data-source
        return dataSources[0];
    };
    return RestClient;
}(errorListener_1.ErrorListener));
exports.RestClient = RestClient;
