// @linkurious/rest-client/dist/src/api/Search/index.js
"use strict";
/**
 * Copyright Linkurious SAS 2012 - 2019
 *
 * - Created on 2019-10-30.
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var request_1 = require("../../http/request");
var response_1 = require("../../http/response");
__export(require("./types"));
var UNAUTHORIZED = response_1.LkErrorKey.UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE = response_1.LkErrorKey.DATA_SOURCE_UNAVAILABLE, GUEST_DISABLED = response_1.LkErrorKey.GUEST_DISABLED, FORBIDDEN = response_1.LkErrorKey.FORBIDDEN, ILLEGAL_SOURCE_STATE = response_1.LkErrorKey.ILLEGAL_SOURCE_STATE, SOURCE_ACTION_NEEDED = response_1.LkErrorKey.SOURCE_ACTION_NEEDED, MALFORMED_SEARCH_SYNTAX = response_1.LkErrorKey.MALFORMED_SEARCH_SYNTAX;
var SearchAPI = /** @class */ (function (_super) {
    __extends(SearchAPI, _super);
    function SearchAPI() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Start the indexation.
     * The API doesn't wait for the indexation to finish.
     */
    SearchAPI.prototype.startIndexation = function (params) {
        return this.request({
            errors: [
                UNAUTHORIZED,
                DATA_SOURCE_UNAVAILABLE,
                FORBIDDEN,
                ILLEGAL_SOURCE_STATE,
                SOURCE_ACTION_NEEDED
            ],
            url: '/:sourceKey/search/index',
            method: 'POST',
            params: params
        });
    };
    /**
     * Get the indexation status for a given data-source.
     */
    SearchAPI.prototype.getIndexationStatus = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, GUEST_DISABLED],
            url: '/:sourceKey/search/status',
            method: 'GET',
            params: params
        });
    };
    /**
     * Perform a search of nodes or edges based on a search query, a fuzziness value and filters.
     * The list of items that matched the search query is returned.
     */
    SearchAPI.prototype.search = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, GUEST_DISABLED, MALFORMED_SEARCH_SYNTAX],
            url: '/:sourceKey/search/:type',
            method: 'POST',
            params: params
        });
    };
    /**
     * Perform a search of nodes or edges based on a search query, a fuzziness value and filters.
     * A subgraph made of the items that matched the search query and the edges between them is returned.
     */
    SearchAPI.prototype.searchFull = function (params) {
        return this.request({
            errors: [UNAUTHORIZED, DATA_SOURCE_UNAVAILABLE, GUEST_DISABLED, MALFORMED_SEARCH_SYNTAX],
            url: '/:sourceKey/search/:type/full',
            method: 'POST',
            params: params
        });
    };
    return SearchAPI;
}(request_1.Request));
exports.SearchAPI = SearchAPI;
