// @linkurious/rest-client/dist/src/api/Search/types.js
"use strict";
/**
 * Copyright Linkurious SAS 2012 - 2019
 *
 * - Created on 2019-10-30.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var IndexationStatus;
(function (IndexationStatus) {
    IndexationStatus["ONGOING"] = "ongoing";
    IndexationStatus["DONE"] = "done";
    IndexationStatus["NEEDED"] = "needed";
})(IndexationStatus = exports.IndexationStatus || (exports.IndexationStatus = {}));
var SearchSyntaxErrorKey;
(function (SearchSyntaxErrorKey) {
    /**
     * 1. Detectable errors by checking one statement
     */
    // Fuzzy-statement:
    // When fuzzy value is not an integer between 1 and 100
    SearchSyntaxErrorKey["INVALID_FUZZINESS"] = "invalid-fuzziness";
    // When there are more than two fuzzy statements /
    SearchSyntaxErrorKey["SEVERAL_FUZZINESS"] = "several-fuzziness";
    // Scope-statement:
    // When scope value is not in SearchQueryScope
    SearchSyntaxErrorKey["INVALID_SCOPE"] = "invalid-scope";
    // When scope value already exists and is different
    SearchSyntaxErrorKey["CONFLICTING_SCOPES"] = "conflicting-scopes";
    // When either nodes or edges are not searchable
    SearchSyntaxErrorKey["EDGES_NOT_SEARCHABLE"] = "edges-not-searchable";
    // Type-statement:
    // When type value already exists and is different
    SearchSyntaxErrorKey["INCOMPATIBLE_TYPE_STATEMENTS"] = "incompatible-type-statements";
    // Property-statement's value
    // Unsupported operator (not thrown in validator but in vendor)
    SearchSyntaxErrorKey["UNSUPPORTED_OPERATOR"] = "unsupported-operator";
    /**
     * 2. Detectable errors by checking two statements
     */
    // Type-statement's value (depends on scope-statement)
    // When node category is not searchable
    SearchSyntaxErrorKey["NODE_TYPE_NOT_SEARCHABLE"] = "node-type-not-searchable";
    // When edge type is not searchable
    SearchSyntaxErrorKey["EDGE_TYPE_NOT_SEARCHABLE"] = "edge-type-not-searchable";
    // Property-statement (depends on scope-statement and type-statement)
    // When property is not searchable or visible in schema with/without itemType
    SearchSyntaxErrorKey["PROPERTIES_NOT_SEARCHABLE"] = "properties-not-searchable";
    // When comparison operator is used with a property that is not a number (or date, and same for ranges)
    SearchSyntaxErrorKey["COMPARATOR_TYPE_MISMATCH"] = "comparator-type-mismatch";
    // When comparison operator with invalid number (or date, and same for ranges)
    SearchSyntaxErrorKey["COMPARATOR_WITH_STRING"] = "comparator-with-string";
    // SearchQuery
    // When there is no term or phrase, and there's at least another statement
    SearchSyntaxErrorKey["EMPTY_SEARCH"] = "empty-search";
    /**
     * 3. Any other error
     */
    // When the grammar is invalid or any other unexpected error
    SearchSyntaxErrorKey["SYNTAX_ERROR"] = "syntax-error";
})(SearchSyntaxErrorKey = exports.SearchSyntaxErrorKey || (exports.SearchSyntaxErrorKey = {}));
